File: packages/java/src/main/java/dev/uncaught/UncaughtClient.java
// ---------------------------------------------------------------------------
// dev.uncaught — UncaughtClient (SDK entry-point)
// ---------------------------------------------------------------------------

package dev.uncaught;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.time.Instant;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.UUID;
import java.util.function.Function;
import java.util.regex.Pattern;

/**
 * Main client for the Uncaught error monitoring SDK.
 * <p>
 * Captures errors through a pipeline:
 * normalise error -> generate fingerprint -> rate-limit -> collect breadcrumbs
 * -> detect environment -> sanitise -> build fix prompt -> transport.send
 * <p>
 * Thread-safe: can be shared across threads.
 */
public class UncaughtClient {

    private static final String SDK_NAME = "dev.uncaught/java";
    private static final String SDK_VERSION = "0.1.0";

    private final Config config;
    private final BreadcrumbStore breadcrumbs;
    private final LocalFileTransport transport;
    private final RateLimiter rateLimiter;
    private final String sessionId;
    private final Set<String> seenFingerprints = new HashSet<>();

    private volatile Types.UserInfo user;

    // -----------------------------------------------------------------------
    // Constructor
    // -----------------------------------------------------------------------

    public UncaughtClient(Config config) {
        this.config = config != null ? config : new Config();
        this.breadcrumbs = new BreadcrumbStore(this.config.getMaxBreadcrumbs());
        this.rateLimiter = new RateLimiter(this.config.getMaxEventsPerMinute());
        this.sessionId = UUID.randomUUID().toString();

        // Determine transport output directory
        String outputDir = this.config.getLocalOutputDir();
        if (outputDir != null && !outputDir.isEmpty()) {
            this.transport = new LocalFileTransport(outputDir);
        } else {
            this.transport = new LocalFileTransport();
        }
    }

    // -----------------------------------------------------------------------
    // Public API
    // -----------------------------------------------------------------------

    /**
     * Return the current SDK configuration.
     */
    public Config getConfig() {
        return config;
    }

    /**
     * Capture a {@link Throwable} and send it through the transport pipeline.
     *
     * @param error   the throwable to capture
     * @param context optional context (request, operation, severity)
     * @return the event ID, or {@code null} if the event was dropped
     */
    public String captureError(Throwable error, CaptureContext context) {
        try {
            if (!config.isEnabled()) return null;

            // --- Normalise error ---
            Types.ErrorInfo errorInfo = normaliseError(error);

            if (context != null && context.getComponentStack() != null) {
                errorInfo.setComponentStack(context.getComponentStack());
            }

            // --- Check ignoreErrors ---
            if (shouldIgnore(errorInfo.getMessage())) {
                debugLog("Event ignored by ignoreErrors filter");
                return null;
            }

            // --- Fingerprint ---
            String fingerprint = Fingerprint.generate(
                    errorInfo.getType(),
                    errorInfo.getMessage(),
                    errorInfo.getStack()
            );

            // --- Rate limit ---
            if (!rateLimiter.shouldAllow(fingerprint)) {
                debugLog("Rate-limited: " + fingerprint);
                return null;
            }

            // --- Collect breadcrumbs ---
            List<Types.Breadcrumb> crumbs = breadcrumbs.getAll();

            // --- Detect environment ---
            Types.EnvironmentInfo environment = EnvironmentDetector.detect();

            // Attach deployment environment from config
            if (config.getEnvironment() != null) {
                // Create a copy to avoid mutating the cached instance
                Types.EnvironmentInfo envCopy = new Types.EnvironmentInfo();
                envCopy.setFramework(environment.getFramework());
                envCopy.setFrameworkVersion(environment.getFrameworkVersion());
                envCopy.setRuntime(environment.getRuntime());
                envCopy.setRuntimeVersion(environment.getRuntimeVersion());
                envCopy.setPlatform(environment.getPlatform());
                envCopy.setOs(environment.getOs());
                envCopy.setLocale(environment.getLocale());
                envCopy.setTimezone(environment.getTimezone());
                envCopy.setDeploy(config.getEnvironment());
                environment = envCopy;
            }

            // --- Build event ---
            String eventId = UUID.randomUUID().toString();
            Types.UncaughtEvent event = new Types.UncaughtEvent();
            event.setEventId(eventId);
            event.setTimestamp(Instant.now().toString());
            event.setProjectKey(config.getProjectKey());
            event.setLevel(context != null && context.getLevel() != null
                    ? context.getLevel()
                    : Types.LEVEL_ERROR);
            event.setFingerprint(fingerprint);
            event.setRelease(config.getRelease());
            event.setError(errorInfo);
            event.setBreadcrumbs(crumbs);
            event.setRequest(context != null ? context.getRequest() : null);
            event.setOperation(context != null ? context.getOperation() : null);
            event.setEnvironment(environment);

            // User info
            Types.UserInfo currentUser = this.user;
            if (currentUser != null) {
                Types.UserInfo userCopy = currentUser.copy();
                userCopy.putExtra("sessionId", sessionId);
                event.setUser(userCopy);
            } else {
                Types.UserInfo anon = new Types.UserInfo();
                anon.putExtra("sessionId", sessionId);
                event.setUser(anon);
            }

            event.setFixPrompt(""); // will be set below
            event.setSdk(new Types.SdkInfo(SDK_NAME, SDK_VERSION));

            // --- Sanitise ---
            event = Sanitizer.sanitize(event, config.getSanitizeKeys());

            // --- Build fix prompt ---
            event.setFixPrompt(PromptBuilder.buildFixPrompt(event));

            // --- beforeSend hook ---
            Function<Types.UncaughtEvent, Types.UncaughtEvent> beforeSend = config.getBeforeSend();
            if (beforeSend != null) {
                Types.UncaughtEvent result = beforeSend.apply(event);
                if (result == null) {
                    debugLog("Event dropped by beforeSend");
                    return null;
                }
                event = result;
            }

            // --- Send ---
            transport.send(event);
            debugLog("Captured event: " + eventId + " (" + fingerprint + ")");

            // Track seen fingerprints
            synchronized (seenFingerprints) {
                seenFingerprints.add(fingerprint);
            }

            return eventId;

        } catch (Exception e) {
            debugLog("captureError failed: " + e.getMessage());
            return null;
        }
    }

    /**
     * Capture a {@link Throwable} with default context.
     */
    public String captureError(Throwable error) {
        return captureError(error, null);
    }

    /**
     * Capture a plain message (not backed by a Throwable).
     */
    public String captureMessage(String message, String level) {
        try {
            return captureError(new Exception(message),
                    new CaptureContext().level(level != null ? level : Types.LEVEL_INFO));
        } catch (Exception e) {
            debugLog("captureMessage failed: " + e.getMessage());
            return null;
        }
    }

    public String captureMessage(String message) {
        return captureMessage(message, Types.LEVEL_INFO);
    }

    /**
     * Add a breadcrumb to the ring buffer.
     */
    public void addBreadcrumb(Types.Breadcrumb crumb) {
        try {
            if (!config.isEnabled()) return;
            breadcrumbs.add(crumb);
        } catch (Exception e) {
            debugLog("addBreadcrumb failed: " + e.getMessage());
        }
    }

    /**
     * Convenience method to add a breadcrumb.
     */
    public void addBreadcrumb(String type, String category, String message) {
        addBreadcrumb(new Types.Breadcrumb(type, category, message));
    }

    /**
     * Set user context that will be attached to subsequent events.
     */
    public void setUser(Types.UserInfo user) {
        this.user = user != null ? user.copy() : null;
    }

    /**
     * Clear user context.
     */
    public void clearUser() {
        this.user = null;
    }

    // -----------------------------------------------------------------------
    // Error normalisation
    // -----------------------------------------------------------------------

    private static Types.ErrorInfo normaliseError(Throwable error) {
        Types.ErrorInfo info = new Types.ErrorInfo();

        if (error == null) {
            info.setMessage("Unknown error");
            info.setType("UnknownError");
            return info;
        }

        info.setMessage(error.getMessage() != null ? error.getMessage() : error.toString());
        info.setType(error.getClass().getSimpleName());

        // Convert stack trace to string
        StringWriter sw = new StringWriter();
        error.printStackTrace(new PrintWriter(sw));
        info.setStack(sw.toString());

        return info;
    }

    // -----------------------------------------------------------------------
    // Internal helpers
    // -----------------------------------------------------------------------

    private boolean shouldIgnore(String message) {
        if (message == null) return false;
        List<Pattern> patterns = config.getIgnoreErrors();
        if (patterns == null || patterns.isEmpty()) return false;

        for (Pattern pattern : patterns) {
            if (pattern.matcher(message).find()) return true;
        }

        return false;
    }

    private void debugLog(String msg) {
        if (config.isDebug()) {
            System.err.println("[uncaught] " + msg);
        }
    }

    // -----------------------------------------------------------------------
    // CaptureContext — optional context for captureError
    // -----------------------------------------------------------------------

    /**
     * Optional context to attach to a captured error.
     */
    public static class CaptureContext {
        private Types.RequestInfo request;
        private Types.OperationInfo operation;
        private String componentStack;
        private String level;

        public CaptureContext request(Types.RequestInfo request) {
            this.request = request;
            return this;
        }

        public CaptureContext operation(Types.OperationInfo operation) {
            this.operation = operation;
            return this;
        }

        public CaptureContext componentStack(String componentStack) {
            this.componentStack = componentStack;
            return this;
        }

        public CaptureContext level(String level) {
            this.level = level;
            return this;
        }

        public Types.RequestInfo getRequest() { return request; }
        public Types.OperationInfo getOperation() { return operation; }
        public String getComponentStack() { return componentStack; }
        public String getLevel() { return level; }
    }
}

File: packages/dotnet/src/Uncaught/UncaughtClient.cs
// ---------------------------------------------------------------------------
// Uncaught — UncaughtClient (SDK entry-point for .NET)
// ---------------------------------------------------------------------------

using System.Text.Json;

namespace Uncaught;

/// <summary>
/// Main client for the Uncaught error monitoring SDK.
/// Thread-safe singleton pattern.
/// </summary>
public sealed class UncaughtClient
{
    private const string SdkName = "uncaught-dotnet";
    private const string SdkVersion = "0.1.0";

    private static UncaughtClient? _instance;
    private static readonly object _lock = new();

    private readonly UncaughtConfig _config;
    private readonly BreadcrumbStore _breadcrumbs;
    private readonly LocalFileTransport _transport;
    private readonly RateLimiter _rateLimiter;
    private readonly string _sessionId;
    private readonly HashSet<string> _seenFingerprints = new();
    private UserInfo? _user;

    /// <summary>
    /// Initialise the Uncaught SDK. Calling this more than once replaces
    /// the previous client instance.
    /// </summary>
    public static UncaughtClient Init(UncaughtConfig? config = null)
    {
        lock (_lock)
        {
            _instance = new UncaughtClient(config ?? new UncaughtConfig());
            return _instance;
        }
    }

    /// <summary>
    /// Return the current singleton client, or null if Init() has not been called.
    /// </summary>
    public static UncaughtClient? GetClient()
    {
        lock (_lock)
        {
            return _instance;
        }
    }

    private UncaughtClient(UncaughtConfig config)
    {
        _config = config;
        _breadcrumbs = new BreadcrumbStore(config.MaxBreadcrumbs);
        _transport = new LocalFileTransport(config);
        _rateLimiter = new RateLimiter(config.MaxEventsPerMinute, 5);
        _sessionId = Guid.NewGuid().ToString();
    }

    /// <summary>
    /// Return the current SDK configuration.
    /// </summary>
    public UncaughtConfig GetConfig() => _config;

    /// <summary>
    /// Capture an exception and send it through the transport pipeline.
    /// Returns the event ID, or null if the event was dropped.
    /// </summary>
    public string? CaptureException(Exception exception, CaptureContext? context = null)
    {
        try
        {
            if (!_config.Enabled) return null;

            var ctx = context ?? new CaptureContext();

            // Normalise error
            var errorInfo = NormaliseException(exception);

            // Check ignoreErrors
            if (ShouldIgnore(errorInfo.Message))
            {
                DebugLog("Event ignored by ignoreErrors filter");
                return null;
            }

            // Generate fingerprint
            var fingerprint = Fingerprint.Generate(
                errorInfo.Type, errorInfo.Message, errorInfo.Stack);

            // Rate limit
            if (!_rateLimiter.ShouldAllow(fingerprint))
            {
                DebugLog($"Rate-limited: {fingerprint}");
                return null;
            }

            // Collect breadcrumbs
            var crumbs = _breadcrumbs.GetAll();

            // Detect environment
            var environment = EnvironmentDetector.Detect();
            if (_config.Environment != null)
            {
                environment.Deploy = _config.Environment;
            }

            // Build user info
            var user = _user != null
                ? new UserInfo
                {
                    Id = _user.Id,
                    Email = _user.Email,
                    Username = _user.Username,
                    SessionId = _sessionId
                }
                : new UserInfo { SessionId = _sessionId };

            // Build event
            var eventId = Guid.NewGuid().ToString();
            var uncaughtEvent = new UncaughtEvent
            {
                EventId = eventId,
                Timestamp = DateTime.UtcNow.ToString("o"),
                ProjectKey = _config.ProjectKey,
                Level = ctx.Level ?? SeverityLevel.Error,
                Fingerprint = fingerprint,
                Release = _config.Release,
                Error = errorInfo,
                Breadcrumbs = crumbs,
                Request = ctx.Request,
                Operation = ctx.Operation,
                Environment = environment,
                User = user,
                FixPrompt = "",
                Sdk = new SdkInfo { Name = SdkName, Version = SdkVersion }
            };

            // Sanitize
            uncaughtEvent = Sanitizer.Sanitize(uncaughtEvent, _config.SanitizeKeys);

            // Build fix prompt
            uncaughtEvent.FixPrompt = PromptBuilder.Build(uncaughtEvent);

            // beforeSend hook
            if (_config.BeforeSend != null)
            {
                var result = _config.BeforeSend(uncaughtEvent);
                if (result == null)
                {
                    DebugLog("Event dropped by beforeSend");
                    return null;
                }
                uncaughtEvent = result;
            }

            // Send
            _transport.Send(uncaughtEvent);
            DebugLog($"Captured event: {eventId} ({fingerprint})");

            // Track seen fingerprints
            lock (_seenFingerprints)
            {
                _seenFingerprints.Add(fingerprint);
            }

            return eventId;
        }
        catch (Exception ex)
        {
            DebugLog($"CaptureException failed: {ex.Message}");
            return null;
        }
    }

    /// <summary>
    /// Capture a plain message.
    /// </summary>
    public string? CaptureMessage(string message, SeverityLevel level = SeverityLevel.Info)
    {
        try
        {
            return CaptureException(
                new Exception(message),
                new CaptureContext { Level = level });
        }
        catch (Exception ex)
        {
            DebugLog($"CaptureMessage failed: {ex.Message}");
            return null;
        }
    }

    /// <summary>
    /// Add a breadcrumb to the ring buffer.
    /// </summary>
    public void AddBreadcrumb(
        BreadcrumbType type,
        string category,
        string message,
        Dictionary<string, object>? data = null,
        SeverityLevel? level = null)
    {
        if (!_config.Enabled) return;

        _breadcrumbs.Add(new Breadcrumb
        {
            Type = type,
            Category = category,
            Message = message,
            Timestamp = DateTime.UtcNow.ToString("o"),
            Data = data,
            Level = level
        });
    }

    /// <summary>
    /// Set user context.
    /// </summary>
    public void SetUser(UserInfo? user) => _user = user;

    /// <summary>
    /// Flush all queued events.
    /// </summary>
    public void Flush() => _transport.Flush();

    // -----------------------------------------------------------------------
    // Internal helpers
    // -----------------------------------------------------------------------

    private static ErrorInfo NormaliseException(Exception ex)
    {
        return new ErrorInfo
        {
            Message = ex.Message ?? ex.ToString(),
            Type = ex.GetType().Name,
            Stack = ex.StackTrace
        };
    }

    private bool ShouldIgnore(string message)
    {
        foreach (var pattern in _config.IgnoreErrors)
        {
            if (message.Contains(pattern, StringComparison.OrdinalIgnoreCase))
            {
                return true;
            }
        }
        return false;
    }

    private void DebugLog(string msg)
    {
        if (_config.Debug)
        {
            Console.Error.WriteLine($"[uncaught] {msg}");
        }
    }
}
